### PR TITLE
Move examples.

### PR DESCRIPTION
**Pull request by miskcoo**

### CODE REVIEW

File: examples/semi_supervised_vae/vae_ssl_rws.py [hunk 1]
```diff
@@ -117,11 +117,18 @@ def log_joint(observed):
...
-        {'z': [qz_samples, log_qz]}, axis=0)
-    labeled_cost = tf.reduce_mean(labeled_cost)
-    labeled_log_likelihood = tf.reduce_mean(labeled_log_likelihood)
+
+    labeled_klpq_obj = zs.variational.klpq(log_joint,
+                                           observed={'x': x_labeled_obs,
+                                                     'y': y_labeled_obs},
+                                           latent={'z': [qz_samples, log_qz]},
+                                           axis=0)
+    labeled_variational_cost = tf.reduce_mean(labeled_klpq_obj.rws())
```
Comment (thjashin) [inline examples/semi_supervised_vae/vae_ssl_rws.py]: Maybe two comments to show which is learning model parameters and which is adapting the proposal.

File: examples/semi_supervised_vae/vae_ssl_rws.py [hunk 2]
```diff
@@ -117,11 +117,18 @@ def log_joint(observed):
...
+    labeled_lower_bound = zs.variational.importance_weighted_objective(
+        log_joint, observed={'x': x_labeled_obs, 'y': y_labeled_obs},
+        latent={'z': [qz_samples, log_qz]}, axis=0)
+    labeled_model_cost = tf.reduce_mean(labeled_lower_bound.sgvb())
```
Comment (thjashin) [inline examples/semi_supervised_vae/vae_ssl_rws.py]: This doesn't need for estimator because are optimizing wrt. model parameters. See doc string.

File: examples/semi_supervised_vae/vae_ssl_rws.py [hunk 3]
```diff
@@ -130,11 +137,18 @@ def log_joint(observed):
...
-        log_joint, {'x': x_unlabeled_obs},
-        {'y': [qy_samples, log_qy], 'z': [qz_samples, log_qz]}, axis=0)
-    unlabeled_cost = tf.reduce_mean(unlabeled_cost)
-    unlabeled_log_likelihood = tf.reduce_mean(unlabeled_log_likelihood)
+
+    unlabeled_klpq_obj = zs.variational.klpq(
+        log_joint, observed={'x': x_unlabeled_obs},
+        latent={'y': [qy_samples, log_qy],
+                'z': [qz_samples, log_qz]}, axis=0)
+    unlabeled_variational_cost = tf.reduce_mean(unlabeled_klpq_obj.rws())
```
Comment (thjashin) [inline examples/semi_supervised_vae/vae_ssl_rws.py]: klpq_cost is better

Comment (thjashin) [reply examples/semi_supervised_vae/vae_ssl_rws.py]: Have you tested the program to check whether the results in first epoches doesn't change too much?

File: zhusuan/variational/base.py [hunk 4]
```diff
@@ -49,16 +49,24 @@ def __init__(self, log_joint, observed, latent):
...
+                          frozenset(six.iteritems(observed)))
+
+    # Cache for log_p and log_q
+    _log_p, _log_q = {}, {}
 
     def _log_joint_term(self):
-        if not hasattr(self, '_log_p'):
-            self._log_p = self._log_joint(self._joint_obs)
-        return self._log_p
+        if self._dict_key not in self._log_p:
```
Comment (thjashin) [inline zhusuan/variational/base.py]: if the cache is in class, should it still be accessed by self? I guess you can check the Context class that BayesianNet is inherited from. maybe some classmethod like _get_cache(), _set_cache() is good? And call them by type(self)._set_cache() or things like this.

File: examples/semi_supervised_vae/vae_ssl_rws.py (hunk 3 above)
Comment (miskcoo) [reply examples/semi_supervised_vae/vae_ssl_rws.py]: Yes, 

Old Version, 
>Epoch 1 (24.7s), Lower bound: labeled = -169.6367950439453, unlabeled = -176.50120544433594 Accuracy: 98.96%
>Epoch 2 (24.2s), Lower bound: labeled = -107.6112060546875, unlabeled = -139.37413024902344 Accuracy: 100.00%

Current Version,
>Epoch 1 (25.0s), Lower bound: labeled = -169.51016235351562, unlabeled = -176.43612670898438 Accuracy: 99.18%
>Epoch 2 (24.6s), Lower bound: labeled = -107.60102844238281, unlabeled = -139.04464721679688 Accuracy: 100.00%

File: zhusuan/variational/base.py (hunk 4 above)
Comment (miskcoo) [reply zhusuan/variational/base.py]: maybe just use classmethod to get _log_p is better? like the last commit?
